client/relay-chain-rpc-interface/src/light_client_worker.rs

```diff
@@ -0,0 +1,290 @@
+		TransportSenderT,
+	},
+	Error,
+};
+use smoldot_light::{ChainId, Client as SmoldotClient, ClientConfig, JsonRpcResponses};
+use std::sync::Arc;
+use tokio::sync::mpsc::{channel as tokio_channel, Receiver, Sender as TokioSender};
+
+use cumulus_primitives_core::relay_chain::{
+	Block as RelayBlock, BlockNumber as RelayNumber, Hash as RelayHash, Header as RelayHeader,
+};
+use cumulus_relay_chain_interface::{RelayChainError, RelayChainResult};
+
+use polkadot_service::generic::SignedBlock;
+
+use sc_rpc_api::chain::ChainApiClient;
+use sc_service::SpawnTaskHandle;
+
+use crate::{rpc_client::RpcDispatcherMessage, tokio_platform::TokioPlatform};
+
+const LOG_TARGET: &str = "rpc-light-client-worker";
+
+#[derive(thiserror::Error, Debug)]
+enum LightClientError {
+	#[error("Error occured while executing smoldot request: {0}")]
+	SmoldotError(String),
+	#[error("Nothing returned from json_rpc_responses")]
+	EmptyReturn,
+}
+
+/// Sending adapter allowing JsonRpsee to send messages to smoldot
+struct SimpleStringSender {
+	inner: smoldot_light::Client<TokioPlatform, ()>,
+	chain_id: ChainId,
+}
+
+#[async_trait::async_trait]
+impl TransportSenderT for SimpleStringSender {
+	type Error = LightClientError;
+
+	async fn send(&mut self, msg: String) -> Result<(), Self::Error> {
+		self.inner
+			.json_rpc_request(msg, self.chain_id)
+			.map_err(|e| LightClientError::SmoldotError(e.to_string()))
+	}
+}
+
+/// Receiving adapter allowing JsonRpsee to receive messages from smoldot
+struct SimpleStringReceiver {
+	inner: JsonRpcResponses,
+}
+
+#[async_trait::async_trait]
+impl TransportReceiverT for SimpleStringReceiver {
+	type Error = LightClientError;
+
+	async fn receive(&mut self) -> Result<ReceivedMessage, Self::Error> {
+		self.inner
+			.next()
+			.await
+			.map(|message| jsonrpsee::core::client::ReceivedMessage::Text(message))
+			.ok_or(LightClientError::EmptyReturn)
+	}
+}
+
+/// Build a smoldot client and add the specified chain spec to it.
+pub async fn build_smoldot_client(
+	spawner: SpawnTaskHandle,
+	chain_spec: &str,
+) -> RelayChainResult<(SmoldotClient<TokioPlatform, ()>, ChainId, JsonRpcResponses)> {
+	let config = ClientConfig {
+		tasks_spawner: Box::new(move |_, task| {
+			spawner.spawn("cumulus-relay-chain-light-client-task", None, task);
+		}),
+		system_name: "cumulus-relay-chain-light-client".to_string(),
+		system_version: env!("CARGO_PKG_VERSION").to_string(),
+	};
+
+	let mut client: SmoldotClient<TokioPlatform, ()> = SmoldotClient::new(config);
+
+	// Ask the client to connect to a chain.
+	let smoldot_light::AddChainSuccess { chain_id, json_rpc_responses } = client
+		.add_chain(smoldot_light::AddChainConfig {
+			specification: chain_spec,
+			disable_json_rpc: false,
+			potential_relay_chains: core::iter::empty(),
+			database_content: "",
+			user_data: (),
+		})
+		.map_err(|e| RelayChainError::GenericError(e.to_string()))?;
+
+	Ok((client, chain_id, json_rpc_responses.expect("JSON RPC is not disabled; qed")))
```

Comment:
Looking at the other examples in the code base, I think the pattern is that we write the reason why it can not fail. And in this case, JSON RPC is not disabled, which is the reason why it will not fail.